File: 2023/ruby/day1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

def part1(input) = puts(add_numbers(input, false))
def part2(input) = puts(add_numbers(input, true))

def add_numbers(input, part2)
  sum = 0
  File.foreach(input) do |line|
    line = replace_words line if part2

    numbers = []
    line.split('').each do |char|
      numbers.push char.to_i if char =~ /[0-9]/
    end

    sum += numbers[0] * 10 + numbers[-1]
  end

  sum
end

def replace_words(line)
  line
    .gsub('one', 'o1e')
    .gsub('two', 't2o')
    .gsub('three', 't3e')
    .gsub('four', 'f4r')
    .gsub('five', 'f5e')
    .gsub('six', 's6x')
    .gsub('seven', 's7n')
    .gsub('eight', 'e8t')
    .gsub('nine', 'n9e')
end

part2('inputs/1')

File: 2023/rust/src/day1.rs
pub fn main(part: usize) {
    let input = std::fs::read_to_string("inputs/1").unwrap();

    match part {
        0 => task1(input),
        1 => task2(input),
        _ => panic!(),
    }
}

pub fn task2(mut input: String) {
    input = input
        .replace("one", "o1e")
        .replace("two", "t2o")
        .replace("three", "t3e")
        .replace("four", "f4r")
        .replace("five", "f5e")
        .replace("six", "s6x")
        .replace("seven", "s7n")
        .replace("eight", "e8t")
        .replace("nine", "n9e");

    task1(input);
}

pub fn task1(input: String) {
    let mut sum = 0;

    for line in input.lines() {
        sum +=
            char::to_digit(line.chars().find(|&x| x.is_ascii_digit()).unwrap(), 10).unwrap() * 10;
        sum += char::to_digit(line.chars().rfind(|&x| x.is_ascii_digit()).unwrap(), 10).unwrap();
    }
    println!("{}", sum);
}
